pkg/compiler/codegen.go

```diff
@@ -1159,14 +1172,45 @@ func (c *codegen) Visit(node ast.Node) ast.Visitor {
 }
 
 // processDefers emits code for `defer` statements.
+// TRY-related opcodes handle exception as follows:
+// 1. CATCH block is executed only if exception has occured.
+// 2. FINALLY block is always executed, but after catch block.
+// Go `defer` statements are a bit different:
+// 1. `defer` is always executed irregardless of whether an exception has occured.
+// 2. `recover` can or can not handle a possible exception.
+// Thus we use the following approach:
+// 1. Throwed exception is saved in a static field X, static fields Y and is set to true.
+// 2. CATCH and FINALLY blocks are the same, and both contain the same CALLs.
+// 3. In CATCH block we set Y to true and emit default return values if it is the last defer.
+// 4. Execute FINALLY block only if Y is false.
 func (c *codegen) processDefers() {
 	for i := len(c.scope.deferStack) - 1; i >= 0; i-- {
 		stmt := c.scope.deferStack[i]
 		after := c.newLabel()
 		emit.Jmp(c.prog.BinWriter, opcode.ENDTRYL, after)
+
+		c.setLabel(stmt.catchLabel)
+		emit.Opcode(c.prog.BinWriter, opcode.DROP) // drop exception, it will be pushed via `recover()`
+		emit.Int(c.prog.BinWriter, 1)
+		c.emitStoreByIndex(varGlobal, finallyProcessedIndex)
+		ast.Walk(c, stmt.expr)
```

Comment:
So you're emitting this code twice? Why? I think it can be avoided if we're to jump around a little.

Reply:
This is either a `CALL` (<= 5 bytes) or a `PUSHA` (`LDLOC`) + `CALLA` (<= 10 bytes), so the overhead is just 2 opcodes.
I have first tried to reuse the code between blocks, this had lead to a longer implementation (both in Go code and bytecode) which is much more difficult to read and debug.